### DISCUSSION TITLE
In the context of an Items Collection, how do I bind to a command on the view model?

### DISCUSSION THREAD

**Discussion by bsampica** [AvaloniaUI/Avalonia#12719] [General]
So, I think this should be easy, but I just can't work out the syntax.  Here's the setup

```
<UserControl...>
    <Design.DataContext>
          <viewModels:HomePageViewModel />
    </Design.DataContext>


    <ItemsRepeater ItemsSource={Binding ACollectionOnTheViewModel }>
       <ItemsRepeater.ItemTemplate>
          <DataTemplate>
                <Button Command="{Binding ??????}" /> <!-- This needs to be an exposed command on the ViewModel, not in the collection of items -->
          </DataTemplate>
       </ItemsRepeater.ItemTemplate>
</UserControl>

```

I've been round and round on using the TemplatedParent which throws an error about being in a data template, and trying to use the RelativeSource  and ancestors to back up to the ViewModel which is on the UserControl DataContext.

Any help would be appreciated, even if you tell me I'm architecting the solution wrong and I need to rework it to make it work without code stinking' hacks.

**Comment by stevemonaco** [comment]
You need to use some form of binding indirection through a control that contains the VM as its `DataContext`. I'll use the `ElementName` shorthand syntax here, but you could use `RelativeSource` parent instead. You can check out the [docs here](https://docs.avaloniaui.net/docs/next/guides/data-binding/binding-to-controls).

```xml
<ItemsRepeater x:Name="items" ItemsSource={Binding ACollectionOnTheViewModel }>
	<ItemsRepeater.ItemTemplate>
		<DataTemplate>
			<Button Command="{Binding #items.DataContext.SomeCommand}" /> <!-- With ReflectionBinding -->
			<Button Command="{Binding #items.((vm:SomeViewModel)DataContext).SomeCommand}" /> <!-- With CompiledBinding -->
		</DataTemplate>
	</ItemsRepeater.ItemTemplate>
</ItemsRepeater>
```

**Reply by bsampica** [reply]
Thank you....one of the two works in this scenario.  

```
<Button Command="{Binding #items.((vm:SomeViewModel)DataContext).SomeCommand}" /> <!-- With CompiledBinding -->
```
WORKS GREAT!  

```
<Button Command="{Binding #items.DataContext.SomeCommand}" /> <!-- With ReflectionBinding -->
```

The first button in your example produces an error that says 

> Unable to resolve field or property 'NavigateTankDetails' in data context of type 'object'

NavigateTankDetails is the ReactiveCommand on the ViewModel that I'm trying to reference.  The second however does work.  It looks like it's doing a cast?

**Reply by stevemonaco** [reply]
So that means you probably have compiled bindings enabled as a default. In that case, `Binding` falls back to `CompiledBinding`. Remember the `DataContext` here is an `object` unless cast (or via reflection). So if you were to use reflection binding, you need to use it explicitly. eg.

```xml
<Button Command="{ReflectionBinding #items.DataContext.SomeCommand}" />
```

**Comment by timunie** [comment]
fyi: TemplateParent is only for ControlTemplates, where you replace the entire look of a control in your own Style. Using named element would be what I suggest also, or use the `$parent[ItemsPresenter]` syntax

**Comment by bsampica** [comment]
In the past in situations like this, I've broken layer barriers slightly, and developed a ViewDataAdapter that turned a Data Model into a View Data Model, that had all the properties, commands, timers, subscriptions...whatever,  on it the view needed.  Has anyone done something like this in Avalonia?

**Reply by timunie** [reply]
In WPF it was way harder to achive than in Avalonia, so I had such an adapter for WPF but never needed it for Avalonia until today.

**Comment by NickBarton** [comment]
Thanks chaps, I had two nested ItemsControl and the trick to solve the problem is to define an x:Name on the outer one.
I was trying to navigate using $parent, nightmare.
